File: src/main/scala/pl/mjankowski/tests/algs/pl/mjankowski/tests/algs/backtracking/Backtracking.scala
package pl.mjankowski.tests.algs.pl.mjankowski.tests.algs.backtracking

import scala.collection.mutable.ListBuffer

/**
 * Created by mjankowski on 2016-05-01.
 */
abstract class Backtracking[T] {

  def processSolution(sol: ListBuffer[T]): Unit

  def isSolution(sol: ListBuffer[T], iter: Int): Boolean

  def constructCandidates(iter: Int): List[T]

  def backtrack(sol: ListBuffer[T], k: Int): Unit = {

    if (isSolution(sol, k)) {
      processSolution(sol)
    } else {
      val candidates = constructCandidates(k + 1)
      for (c <- candidates) {
        sol += c
        backtrack(sol, k + 1)
        sol.remove(sol.length - 1)
      }
    }
  }
}

File: src/main/java/pl/mjankowski/tests/algs/Backtracking.java
package pl.mjankowski.tests.algs;

import java.util.List;

/**
 * Created with IntelliJ IDEA.
 * User: mac
 * Date: 03.05.16
 * Time: 21:49
 * To change this template use File | Settings | File Templates.
 */
public abstract class Backtracking<T> {

    public void backtrack(List<T> solution, int iteration) {

        if (isSolution(solution)) {
            processSolution(solution);
        } else {

            if (isAllowed(solution, iteration)) {
//                printDebugSolution(solution, true);
                List<T> candidates = generateCandidates(solution, iteration);
                for (T c : candidates) {
                    solution.add(c);
                    backtrack(solution, iteration + 1);
                    solution.remove(solution.size() - 1);
                }
            }else{
//                printDebugSolution(solution, false);
            }
        }
    }

    protected abstract boolean isSolution(List<T> solution);

    protected abstract void processSolution(List<T> solution);

    protected abstract List<T> generateCandidates(List<T> current, int iteration);

    protected abstract boolean isAllowed(List<T> solution, int iteration);

    protected abstract void printDebugSolution(List<T> solution, boolean allowed);

}
